Modules/JavaScript/Classes-Constructor-New/homework.js

```diff
@@ -0,0 +1,86 @@
+console.log(`
+  Task.
+
+    Create a Vector object that supports:
+     1. add (same length)
+     2. subtract
+     3. dot
+     4. norm
+     5. If you try to add, subtract, or dot two vectors with different lengths, you must throw an error.
+     6. toString
+     7. equals
+`);
+
+class Vector {
+  list = [];
+
+  constructor(list) {
+    this.list = list;
+  }
+
+  add(vector) {
+    this.checkIfVectorsLengthsMatch(vector);
+    
+    return new Vector(this.list.map((val, index) => val + vector.list[index]));
+  }
+  
+  subtract(vector) {
+    this.checkIfVectorsLengthsMatch(vector);
+    
+    return new Vector(this.list.map((val, index) => val - vector.list[index]));
+  }
+  
+  dot(vector) {
+    this.checkIfVectorsLengthsMatch(vector);
+    
+    return this.list.reduce((sum, val, index) => {
+      console.log('  ----- val=', val, 'vector.list[index]=', vector.list[index], 'addThis=', (val * vector.list[index]))
+      return sum + (val * vector.list[index]);
+    }, 0)
+  }
+  
+  norm() {
+    const sum = this.list.reduce((a, b) => {
+      console.log('  ----- val*2=', b * 2);
+      return a + (b * 2);
+    }, 0);
+    return Math.sqrt(sum);
+  }
+
+  toString() {
+    return `(${this.list.toString()})`;
+  }
+
+  equals(vector) {
+    if (this.list.length == vector.list.length) {
```

Comment:
why not use your already defined validator?
also, please try to avoid == in favor of === to avoid unexpected type conversions

Reply:
could use `===` , but here it would always result in `number` type, so don't think that it's actual needed.

about validator - then it would be wrapped in try/catch block which doesn't look nice

Reply:
You should definitely get used to always using `===`

About validator, why don't you return true or false, as the name of the validator function states? It is not called `throwIfInvalid` or something like that, right? This would make it reusable